### PR TITLE
ISSUE-298 Implement CassandraRateLimitingPlanRepository

### PR DESCRIPTION
**Pull request by vttranlina**
Ref https://github.com/linagora/james-project-private/issues/298

### CODE REVIEW

**Comment by vttranlina** [comment]
I'm not sure about this way is best

## Current way: Each row in the table will store each entry. 
```scala
  case class RateLimitingPlanEntry(planId: UUID,
                                   planName: String,
                                   operationLimitation: String,
                                   rateLimitationName: String,
                                   period: Long,
                                   limitType: String,
                                   allowedQuantity: Long)
```
### Cons: 
- duplicate data in multi columns (planId, planName, operationLimitation, rateLimitationName, period)
- more boilerplate

### Pros
- Query faster?
- No need serialize/deserialize data
- We have a small number of RateLimitingPlan, so duplicate data is not a problem?

## Another way
1. Create another table, and reference between each table
Cons:
- more table

2. Store all data of `operationLimitations` in a single column. (maybe JSON format?)
Cons: 
- I think it will not be acceptable. In past, we avoid its pattern

**Comment by quantranhong1999** [comment]
> duplicate data in multi columns (planId, planName, operationLimitation, rateLimitationName, period)

I remember Cassandra having a mechanism to compact the data so I don't think this is a big concern.

**Comment by chibenwa** [comment]
Please invastigate User defined types and MAP datastructure. Beware of the updates (can we freeze types?). I'd like to have only one table if possible...

I'll try to get a closer look tomorrow morning.

**Comment by Arsnael** [comment]
Is it really that much duplication of data though? Rereading the tickets, I'm not convinced... I would say that your planName likely is duplicated, but after the others can be different?

**Comment by vttranlina** [comment]
> Is it really that much duplication of data though? Rereading the tickets, I'm not convinced... I would say that your planName likely is duplicated, but after the others can be different?

Source code has a bit different from what's written in the ticket. 
Some type has been abstract.
https://github.com/linagora/tmail-backend/blob/b5eb384888b8ef037d7a7be5a11da368f90c3de1/tmail-backend/rate-limiter/rate-limiter-api/src/main/scala/com/linagora/tmail/rate/limiter/api/RateLimitation.scala#L92

**Comment by chibenwa** [comment]
Define a UDT for OperationLimitations

Then have it fit in one row.

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanDAO.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.james.backends.cassandra.init.CassandraTypesProvider
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import java.util.UUID
+import javax.inject.Inject
+import scala.jdk.CollectionConverters._
+
+object CassandraRateLimitPlanDAO {
+  object RateLimitingPlanEntry {
```
Comment (chibenwa) [inline tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanDAO.scala:18]: I have doubts regarding nested object.

Being able to do it do not mean that we should...

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    .statement(statement => statement
+      .addColumn(RateLimitationBase.PERIOD, bigint())
+      .addColumn(RateLimitationBase.LIMIT_MAPPING, frozenMap(text(), bigint())))
+    .table(TABLE_NAME)
+    .comment("Hold Rate Limiting Plan - Use to management.")
+    .statement((statement: Create) => statement
+      .addPartitionKey(PLAN_ID, uuid)
+      .addClusteringColumn(PLAN_NAME, text)
+      .addClusteringColumn(OPERATION_LIMITATION_NAME, text)
+      .addUDTMapColumn(RATE_LIMITATION_MAPPING, text(), SchemaBuilder.frozen(RATE_LIMITATION_BASE)))
```
Comment (chibenwa) [inline tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala:32]: We should likely find a way to freeze this UDT Map too (and not just the UDTs it contains...)

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitationPlanRepository.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      .`then`()
+      .`then`(SMono.just(planId))
+
+  override def update(resetRequest: RateLimitingPlanResetRequest): Publisher[RateLimitingPlan] =
+    SMono.fromPublisher(cassandraRateLimitPlanDAO.planExists(resetRequest.id))
+      .filter(exists => exists)
+      .switchIfEmpty(SMono.error(new RateLimitingPlanNotFoundException))
+      .flatMap(_ => cassandraRateLimitPlanDAO.delete(resetRequest.id))
+      .`then`(create(resetRequest.id, RateLimitingPlanEntry.from(resetRequest)))
+      .`then`(SMono.fromPublisher(get(resetRequest.id)))
```
Comment (chibenwa) [inline tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitationPlanRepository.scala:32]: We don't need another read as we do reset things here

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala (hunk 2 above)
Comment (chibenwa) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala:32]: Eg in CassandraMailQueueViewModule

```
.addColumn(EnqueuedMailsTable.ATTRIBUTES, frozenMap(text(), blob()))
```

(Where blob is a frozen UDT)

If required we could switch UDTs to tuple types.

Eg:

```
.addColumn(EnqueuedMailsTable.PER_RECIPIENT_SPECIFIC_HEADERS, frozenList(TupleType.of(ProtocolVersion.NEWEST_SUPPORTED, CodecRegistry.DEFAULT_INSTANCE, text(), text(), text()))))
```

(good luck!)

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitationPlanRepository.scala (hunk 3 above)
Comment (vttranlina) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitationPlanRepository.scala:32]: read for api 
`def update(resetRequest: RateLimitingPlanResetRequest): Publisher[RateLimitingPlan]`

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanDAO.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanDAO.scala:18]: I just want to notice the purpose of `RateLimitingPlanEntry` is to support `CassandraRateLimitPlanDAO`

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitationPlanRepository.scala (hunk 3 above)
Comment (chibenwa) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitationPlanRepository.scala:32]: Yes but we just read something that we just wrote. We don't need that.

File: tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala:32]: > .addColumn(EnqueuedMailsTable.ATTRIBUTES, frozenMap(text(), blob()))

frozenMap not accept UDTType

> If required we could switch UDTs to tuple types.

```
      .addColumn(RATE_LIMITATIONS,
        frozenList(TupleType.of(ProtocolVersion.NEWEST_SUPPORTED, CodecRegistry.DEFAULT_INSTANCE,
          text(),
          bigint(),
          DataType.map(text, bigint())))))
```

it looks good?

Comment (chibenwa) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala:32]: Better but should'nt we also freeze the map in the tuple?

Comment (chibenwa) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala:32]:
```      .addColumn(RATE_LIMITATIONS,
        frozenList(TupleType.of(ProtocolVersion.NEWEST_SUPPORTED, CodecRegistry.DEFAULT_INSTANCE,
          text(),
          bigint(),
          DataType.frozenMap(text, bigint())))))

```

Comment (chibenwa) [reply tmail-backend/rate-limiter/rate-limiter-cassandra/src/main/scala/com/linagora/tmail/rate/limiter/cassandra/CassandraRateLimitPlanTable.scala:32]: Yes please ensure everything is frozen

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Rebase?

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
@vttranlina Can you rebase and solve conflicts?

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**